junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestInstancePreDestroyCallback.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.extension;
+
+import static org.apiguardian.api.API.Status.STABLE;
+
+import org.apiguardian.api.API;
+
+/**
+ * {@code TestInstancePreDestroyCallback} defines the API for {@link Extension
+ * Extensions} that wish to process test instances <em>after</em> they have been
+ * used in tests and <em>before</em> they are destroyed.
+ *
+ * <p>Common use cases include cleaning dependencies that have been injected into the
+ * test instance, invoking custom de-initialization methods on the test instance, etc.
+ *
+ * <p>Extensions that implement {@code TestInstancePreDestroyCallback} must be
+ * registered at the class level.
+ *
+ * <h3>Constructor Requirements</h3>
+ *
+ * <p>Consult the documentation in {@link Extension} for details on
+ * constructor requirements.
+ *
+ * @since 5.6
+ * @see #preDestroyTestInstance(Object, ExtensionContext)
+ * @see TestInstanceFactory
+ * @see ParameterResolver
+ */
+@FunctionalInterface
+@API(status = STABLE, since = "5.6")
+public interface TestInstancePreDestroyCallback extends Extension {
+
+	/**
+	 * Callback for processing the supplied test instance before it is destroyed.
+	 *
+	 * @param testInstance the instance to process; never {@code null}
+	 * @param context the current extension context; never {@code null}
+	 */
+	void preDestroyTestInstance(Object testInstance, ExtensionContext context) throws Exception;
```

Comment:
Now that I thought a bit more about it you were right, I think you can get rid of the `testInstance` parameter! \o/

Reply:
lets do it!

Reply:
my only comment on this is the signature miss-match between TestInstancePostProcessor and TestInstancePreDestroyCallback, almost feels better to keep the extra parameter for sake of symmetry?

Reply:
I don't think it's necessary. If you look closely to the other similar extensions, they all provide only an `ExtensionContext` with the instance inside it. The only reason why `TestInstancePostProcessor` also has a `testInstance` is explained [in this note](https://github.com/junit-team/junit5/blob/master/junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/TestInstancePostProcessor.java#L45-L49). Since here you already have access to the test instance through the context, I think there's no use keeping `testInstance`?

Reply:
yes, you've put forward a good case to remove the param, so I shall remove it :)